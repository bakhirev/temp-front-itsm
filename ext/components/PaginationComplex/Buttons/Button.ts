import styled, { css } from 'styled-components';

import { DEFAULT_THEME } from '../../common';

const SIZE = 40;

export const Button = styled.button`
  ${({ theme }) => css`
    align-items: center;
    background: ${theme.color.neutral[10]};
    border-radius: ${theme.borderRadius};
    border: none;
    cursor: pointer;
    display: flex;
    height: ${SIZE}px;
    justify-content: center;
    outline: none;
    width: ${SIZE}px;

    & svg {
      fill: ${theme.color.neutral[50]};
    }

    &:focus {
      border: 2px solid ${theme.color.primary[60]};
    }

    &:hover {
      background: ${theme.color.neutral[20]};

      & svg {
        fill: ${theme.color.primary[60]};
      }
    }

    &:active {
      background: ${theme.color.neutral[30]};

      & div svg {
        fill: ${theme.color.primary[60]};
      }
    }

    &:disabled {
      background: ${theme.color.neutral[10]};
      cursor: default;

      & svg {
        fill: ${theme.color.neutral[30]};
      }
    }
  `};
`;

Button.defaultProps = {
  theme: DEFAULT_THEME,
};
