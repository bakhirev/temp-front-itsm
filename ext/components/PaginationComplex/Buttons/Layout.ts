import styled from 'styled-components';

const WIDTH = 88;

export const Layout = styled.div`
  display: flex;
  justify-content: space-between;
  width: ${WIDTH}px;
`;
