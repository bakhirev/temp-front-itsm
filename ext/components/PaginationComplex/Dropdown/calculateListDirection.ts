import type { MutableRefObject } from 'react';

import type { Direction } from './List';

const SAFE_SPACE = 16;

export const calculateListDirection = ({
  dropdownRef,
  listRef,
}: {
  dropdownRef: MutableRefObject<HTMLButtonElement | null>;
  listRef: MutableRefObject<HTMLDivElement | null>;
}): Direction => {
  const dropdownRect = dropdownRef.current?.getBoundingClientRect();
  const listRect = listRef.current?.getBoundingClientRect();
  if (!dropdownRect || !listRect) return 'bottom';

  const { bottom: dropdownBottom } = dropdownRect;
  const { height: listHeight } = listRect;
  const viewPortHeight = window.innerHeight;

  return viewPortHeight - dropdownBottom >= listHeight + SAFE_SPACE ? 'bottom' : 'top';
};
