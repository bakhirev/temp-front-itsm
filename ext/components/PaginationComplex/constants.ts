export const FIRST_PAGE = 1;
export const ICON_SIZE = 20;
