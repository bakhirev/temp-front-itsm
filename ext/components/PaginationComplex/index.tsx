import React, { FC } from 'react';

import { Page } from './Page';
import { PaginationComplexComponent } from './PaginationComplexComponent';
import { Range } from './Range';

const DEFAULT_RANGES = [20, 40, 60, 80];

export interface IPaginationComplexProps {
  /** Имя класса для переопределения стилей */
  className?: string;
  /** data-test-id атрибут для тестирования компонента */
  dataTestId?: string;
  /** Отключить выбор количества элементов на странице через выпадающее меню */
  disableItemsPerPageSelect?: boolean;
  /** Отключить выбор страницы через выпадающее меню */
  disablePageSelect?: boolean;
  /** Количество элементов на странице */
  itemsPerPage: number;
  /** Мобильный вид компонента */
  mobile?: boolean;
  /** Страница для отображения */
  page: number;
  /** Варианты выбора количества элементов на странице */
  ranges?: Array<number>;
  /** Число всех элементов */
  totalItems: number;
  /** Ширина компонента */
  width?: number | string;
  /** Коллбэк на изменение количества элементов на странице */
  onItemsPerPageChange?: (newItemsPerPage: number) => void;
  /** Коллбэк на изменение страницы */
  onPageChange: (newPage: number) => void;
}

export const PaginationComplex: FC<IPaginationComplexProps> = ({
  className,
  dataTestId,
  disableItemsPerPageSelect = false,
  disablePageSelect = false,
  itemsPerPage,
  mobile = false,
  page,
  ranges = DEFAULT_RANGES,
  totalItems,
  width,
  onItemsPerPageChange,
  onPageChange,
}) => {
  const lastPage = Math.ceil(totalItems / itemsPerPage);

  return (
    <PaginationComplexComponent className={className} data-test-id={dataTestId} mobile={mobile}>
      <Range
        disableItemsPerPageSelect={disableItemsPerPageSelect}
        itemsPerPage={itemsPerPage}
        lastPage={lastPage}
        mobile={mobile}
        page={page}
        ranges={ranges}
        totalItems={totalItems}
        onChange={onItemsPerPageChange}
      />
      <Page
        disablePageSelect={disablePageSelect}
        lastPage={lastPage}
        mobile={mobile}
        page={page}
        onChange={onPageChange}
      />
    </PaginationComplexComponent>
  );
};
