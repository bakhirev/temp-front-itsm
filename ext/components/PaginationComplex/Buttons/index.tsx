import React, { FC, MouseEvent } from 'react';

import { ReactComponent as ChevronLeftOutline } from '../../Icons/system/ChevronLeftOutline.svg';
import { ReactComponent as ChevronRightOutline } from '../../Icons/system/ChevronRightOutline.svg';
import { FIRST_PAGE, ICON_SIZE } from '../constants';

import { Button } from './Button';
import { Layout } from './Layout';

interface IButtonProps {
  lastPage: number;
  page: number;
  onClick: (newPage: number) => void;
}

export const Buttons: FC<IButtonProps> = ({ lastPage, page, onClick }) => {
  const handleFocusPrevent = (event: MouseEvent<HTMLButtonElement>) => event.preventDefault();

  const previousPage = page - 1;
  const nextPage = page + 1;

  return (
    <Layout>
      <Button
        disabled={previousPage < FIRST_PAGE}
        onClick={() => onClick(previousPage)}
        onMouseDown={handleFocusPrevent}
      >
        <ChevronLeftOutline width={ICON_SIZE} height={ICON_SIZE} />
      </Button>
      <Button
        disabled={nextPage > lastPage}
        onClick={() => onClick(nextPage)}
        onMouseDown={handleFocusPrevent}
      >
        <ChevronRightOutline width={ICON_SIZE} height={ICON_SIZE} />
      </Button>
    </Layout>
  );
};
