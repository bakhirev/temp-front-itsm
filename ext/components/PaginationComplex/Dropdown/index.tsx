import React, { FC, MouseEvent, useEffect, useRef, useState } from 'react';
import OutsideClickHandler from 'react-outside-click-handler';

import { ReactComponent as ChevronDownOutline } from '../../Icons/system/ChevronDownOutline.svg';
import { ReactComponent as ChevronUpOutline } from '../../Icons/system/ChevronUpOutline.svg';
import { Body2Long } from '../../Typography';
import { KEY_CODES } from '../../common';
import { ICON_SIZE } from '../constants';

import { DropdownComponent } from './DropdownComponent';
import { IconLayout } from './IconLayout';
import { Item } from './Item';
import { List } from './List';
import { calculateListDirection } from './calculateListDirection';
import type { Direction } from './List';

interface IDropdownProps {
  disabled: boolean;
  list: Array<number>;
  value: number;
  onChange?: (newPage: number) => void;
}

export const Dropdown: FC<IDropdownProps> = ({ disabled, list, value, onChange }) => {
  const dropdownRef = useRef<HTMLButtonElement | null>(null);
  const listRef = useRef<HTMLDivElement | null>(null);

  const [opened, setOpened] = useState<boolean>(false);
  const [focusedItem, setFocusedItem] = useState<number>(list[0]);
  const [listDirection, setListDirection] = useState<Direction>('bottom');

  const openList = () => {
    setOpened(true);
    setListDirection(calculateListDirection({ dropdownRef, listRef }));
  };

  const handleFocusPrevent = (event: MouseEvent<HTMLButtonElement>) => event.preventDefault();
  const handleListOpen = () => {
    if (!opened) {
      setFocusedItem(list[0]);
      openList();
    } else {
      setOpened(false);
    }
  };
  const handleOutsideClick = () => setOpened(false);
  const handleEscKeyDown = ({ keyCode }: KeyboardEvent) => {
    opened && keyCode === KEY_CODES.ESCAPE && setOpened(false);
  };
  const handleItemSelect = (item: number) => onChange?.(item);

  const handleItemMouseEnter = (item: number) => setFocusedItem(item);

  const handleArrowDownKeyDown = (event: KeyboardEvent) => {
    if (opened && event.keyCode === KEY_CODES.DOWN) {
      event.preventDefault();
      const focusedItemIdx = list.indexOf(focusedItem);
      const nextFocusedItemIdx = focusedItemIdx + 1;
      if (nextFocusedItemIdx !== list.length) {
        setFocusedItem(list[nextFocusedItemIdx]);
      }
    }
  };

  const handleArrowUpKeyDown = (event: KeyboardEvent) => {
    if (opened && event.keyCode === KEY_CODES.UP) {
      event.preventDefault();
      const focusedItemIdx = list.indexOf(focusedItem);
      const nextFocusedItemIdx = focusedItemIdx - 1;
      if (nextFocusedItemIdx >= 0) {
        setFocusedItem(list[nextFocusedItemIdx]);
      }
    }
  };

  const handleEnterKeyDown = (event: KeyboardEvent) => {
    if (opened && event.keyCode === KEY_CODES.ENTER) {
      onChange?.(focusedItem);
    }
  };

  useEffect(() => {
    document.addEventListener('keydown', handleEscKeyDown);
    document.addEventListener('keydown', handleArrowDownKeyDown);
    document.addEventListener('keydown', handleArrowUpKeyDown);
    document.addEventListener('keydown', handleEnterKeyDown);
    return () => {
      document.removeEventListener('keydown', handleEscKeyDown);
      document.removeEventListener('keydown', handleArrowDownKeyDown);
      document.removeEventListener('keydown', handleArrowUpKeyDown);
      document.removeEventListener('keydown', handleEnterKeyDown);
    };
  });

  const Icon = opened ? ChevronUpOutline : ChevronDownOutline;

  return (
    <OutsideClickHandler onOutsideClick={handleOutsideClick}>
      <DropdownComponent
        disabled={disabled}
        opened={opened}
        ref={dropdownRef}
        onClick={handleListOpen}
        onMouseDown={handleFocusPrevent}
      >
        <Body2Long>{value}</Body2Long>
        <IconLayout>
          <Icon width={ICON_SIZE} height={ICON_SIZE} />
        </IconLayout>
        <List $direction={listDirection} $opened={opened} ref={listRef}>
          {list.map((item, idx) => (
            <Item
              focused={focusedItem === item}
              key={idx}
              selected={value === item}
              onClick={() => handleItemSelect(item)}
              onMouseEnter={() => handleItemMouseEnter(item)}
            >
              <Body2Long>{item}</Body2Long>
            </Item>
          ))}
        </List>
      </DropdownComponent>
    </OutsideClickHandler>
  );
};
