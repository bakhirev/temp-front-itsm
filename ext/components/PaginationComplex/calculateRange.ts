import { FIRST_PAGE } from './constants';

export const calculateRange = ({
  itemsPerPage,
  lastPage,
  page,
  totalItems,
}: {
  itemsPerPage: number;
  lastPage: number;
  page: number;
  totalItems: number;
}) => {
  const calculateStart = () => (page === FIRST_PAGE ? 1 : (page - 1) * itemsPerPage + 1);
  const calculateEnd = () => {
    if (page === FIRST_PAGE) return itemsPerPage;
    if (page === lastPage) return totalItems;
    return itemsPerPage * page;
  };

  return {
    start: calculateStart(),
    end: calculateEnd(),
  };
};
