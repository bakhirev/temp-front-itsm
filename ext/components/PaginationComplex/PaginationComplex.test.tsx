import React, { ReactNode } from 'react';
import { act } from 'react-dom/test-utils';
import { mount, shallow } from 'enzyme';

import { PaginationComplex } from '../PaginationComplex';
import { KEY_CODES } from '../common';

import { Buttons } from './Buttons';
import { Dropdown } from './Dropdown';

jest.mock('../common', () => ({
  DEFAULT_THEME: {
    borderRadius: '',
    boxShadow: {
      8: '',
    },
    color: {
      neutral: {
        10: '',
        20: '',
        30: '',
        50: '',
      },
      primary: {
        60: '',
      },
    },
  },
  KEY_CODES: {
    DOWN: 40,
    ENTER: 13,
  },
}));

jest.mock('../Icons/system/ChevronLeftOutline.svg', () => ({
  // eslint-disable-next-line react/display-name
  ReactComponent: () => <div> {'<'} </div>,
}));
jest.mock('/../Icons/system/ChevronRightOutline.svg', () => ({
  // eslint-disable-next-line react/display-name
  ReactComponent: () => <div> {'>'} </div>,
}));
jest.mock('../Icons/system/ChevronDownOutline.svg', () => ({
  // eslint-disable-next-line react/display-name
  ReactComponent: () => <div> v </div>,
}));
jest.mock('/../Icons/system/ChevronUpOutline.svg', () => ({
  // eslint-disable-next-line react/display-name
  ReactComponent: () => <div> {'/\\'} </div>,
}));

jest.mock('../Typography', () => ({
  // eslint-disable-next-line react/display-name
  Body2Long: ({ children }: { children: ReactNode }) => <div>{children}</div>,
}));

describe('PaginationComplex', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  const arrowDown = new KeyboardEvent('keydown');
  Object.defineProperty(arrowDown, 'keyCode', {
    get: () => KEY_CODES.DOWN,
  });

  const enter = new KeyboardEvent('keydown');
  Object.defineProperty(enter, 'keyCode', {
    get: () => KEY_CODES.ENTER,
  });

  const props = {
    itemsPerPage: 20,
    page: 1,
    totalItems: 113,
    onItemsPerPageChange: jest.fn(),
    onPageChange: jest.fn(),
  };

  it('should render component', () => {
    const wrapper = shallow(<PaginationComplex {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should call onPageChange with next page when user clicks right button', () => {
    const wrapper = shallow(<PaginationComplex {...props} />);
    const rightButton = wrapper.childAt(1).dive().childAt(5).dive().childAt(1);
    rightButton.simulate('click');
    expect(props.onPageChange).toBeCalledWith(props.page + 1);
  });

  it('should call onPageChange with previous page when user clicks left button', () => {
    const page = 2;
    const wrapper = shallow(<PaginationComplex {...props} page={page} />);
    const leftButton = wrapper.childAt(1).dive().childAt(5).dive().childAt(0);
    leftButton.simulate('click');
    expect(props.onPageChange).toBeCalledWith(page - 1);
  });

  it("shouldn't call onPageChange with next page when user clicks right button and component shows last page", () => {
    const page = Math.ceil(props.totalItems / props.itemsPerPage);
    const wrapper = mount(<PaginationComplex {...props} page={page} />);
    const rightButton = wrapper.find(Buttons).find('button').last();
    rightButton.simulate('click');
    expect(props.onPageChange).toBeCalledTimes(0);
  });

  it("shouldn't call onPageChange with previous page when user clicks right button and component shows first page", () => {
    const wrapper = mount(<PaginationComplex {...props} />);
    const leftButton = wrapper.find(Buttons).find('button').first();
    leftButton.simulate('click');
    expect(props.onPageChange).toBeCalledTimes(0);
  });

  it("shouldn't call onPageChange when user clicks buttons and there is only one page", () => {
    const wrapper = mount(<PaginationComplex {...props} totalItems={props.itemsPerPage} />);
    const leftButton = wrapper.find(Buttons).find('button').first();
    const rightButton = wrapper.find(Buttons).find('button').last();
    leftButton.simulate('click');
    rightButton.simulate('click');
    expect(props.onPageChange).toBeCalledTimes(0);
  });

  it('should call onPageChange when user selects page with dropdown', () => {
    const wrapper = mount(<PaginationComplex {...props} />);
    const dropdown = wrapper.find(Dropdown).at(1).find('button');

    dropdown.simulate('click');
    act(() => {
      document.dispatchEvent(arrowDown);
      document.dispatchEvent(enter);
    });

    expect(props.onPageChange).toBeCalledTimes(1);
  });

  it('should call onItemsPerPageChange when user selects range with dropdown', () => {
    const wrapper = mount(<PaginationComplex {...props} />);
    const dropdown = wrapper.find(Dropdown).at(0).find('button');

    dropdown.simulate('click');
    act(() => {
      document.dispatchEvent(arrowDown);
      document.dispatchEvent(enter);
    });

    expect(props.onItemsPerPageChange).toBeCalledTimes(1);
  });
});
