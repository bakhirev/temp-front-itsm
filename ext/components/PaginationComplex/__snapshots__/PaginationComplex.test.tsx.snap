// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaginationComplex should render component 1`] = `
<styled.div
  mobile={false}
>
  <Component
    disableItemsPerPageSelect={false}
    itemsPerPage={20}
    lastPage={6}
    mobile={false}
    onChange={[MockFunction]}
    page={1}
    ranges={
      Array [
        20,
        40,
        60,
        80,
      ]
    }
    totalItems={113}
  />
  <Component
    disablePageSelect={false}
    lastPage={6}
    mobile={false}
    onChange={[MockFunction]}
    page={1}
  />
</styled.div>
`;
